pkg/kv/kvserver/rafttrace/rafttrace.go

```diff
@@ -0,0 +1,459 @@
+// Copyright 2024 The Cockroach Authors.
+//
+// Use of this software is governed by the CockroachDB Software License
+// included in the /LICENSE file.
+
+package rafttrace
+
+import (
+	"context"
+	"math"
+	"sync/atomic"
+
+	"github.com/cockroachdb/cockroach/pkg/kv/kvpb"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvserverpb"
+	"github.com/cockroachdb/cockroach/pkg/raft"
+	"github.com/cockroachdb/cockroach/pkg/raft/raftpb"
+	"github.com/cockroachdb/cockroach/pkg/settings"
+	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+	"github.com/cockroachdb/cockroach/pkg/util/tracing"
+	"github.com/cockroachdb/cockroach/pkg/util/tracing/tracingpb"
+	"github.com/cockroachdb/logtags"
+	"github.com/cockroachdb/redact"
+)
+
+// traceValue represents the trace information for a single registration.
+type traceValue struct {
+	traced kvserverpb.TracedEntry
+	// ctx is a trace specific context used to log events on this trace.
+	ctx context.Context
+
+	mu struct {
+		syncutil.Mutex
+
+		// seenMsgAppResp tracks whether a MsgAppResp message has already been
+		// logged by each replica peer. This limits the size of the log at a
+		// small risk of missing some important messages in the case of dropped
+		// messages or reproposals.
+		seenMsgAppResp map[raftpb.PeerID]bool
+
+		// seenMsgStorageAppendResp tracks whether a MsgStorageAppendResp
+		// message has already been logged.
+		seenMsgStorageAppendResp bool
+
+		// propCtx is the underlying proposal context used for tracing to the
+		// SQL trace.
+		propCtx context.Context
+
+		// propSpan is the span connected to the propCtx. It must be finished
+		// when the trace is removed.
+		propSpan *tracing.Span
+	}
+}
+
+// logf logs the message to the trace context and the proposal context. The
+// proposal context is populated on the leaseholder and is attached to the SQL
+// trace.
+func (t *traceValue) logf(depth int, format string, args ...interface{}) {
+	log.InfofDepth(t.ctx, depth+1, format, args...)
+
+	t.mu.Lock()
+	propCtx := t.mu.propCtx
+	t.mu.Unlock()
+	if propCtx != nil {
+		log.VEventfDepth(propCtx, depth+1, 3, format, args...)
+	}
+}
+
+// sawMsgAppResp returns true on the first MsgAppResp for this peer.
+func (t *traceValue) sawMsgAppResp(p raftpb.PeerID) bool {
```

Comment:
super-nit: rename to `seenMsgAppResp`, to be consistent with the field names. Same for the other "seen" method.

Reply:
Changed